packages/langium-cli/src/generator/highlighting/monarch-generator.ts

```diff
@@ -0,0 +1,476 @@
+
+    const symbols = getSymbols(grammar);
+    const regex = /[{}[\]()]/;
+    const operators = symbols.filter(s => !regex.test(s));
+
+    // build absract monarch grammar representation
+    const monarchGrammar: MonarchGrammar = {
+        languageDefinition: {
+            name:       config.id,  // identifier for generating the grammar export
+            keywords:   getKeywords(grammar),
+            operators,
+            symbols,
+            tokenPostfix: '.' + config.id, // category appended to all tokens
+        },
+        tokenizer: {
+            states: getTokenizerStates(grammar)
+        }
+    };
+
+    // return concrete monarch grammar representation
+    return prettyPrint(monarchGrammar);
+}
+
+/**
+ * Gets Monarch tokenizer states from a Langium grammar
+ * @param grammar Langium grammar to source tokenizer states from
+ * @returns Array of tokenizer states
+ */
+function getTokenizerStates(grammar: langium.Grammar): State[] {
+
+    // initial state, name is arbitrary, just needs to come first
+    const initialState: State = {
+        name: 'initial',
+        rules: getTerminalRules(grammar)
+    };
+
+    const whitespaceState: State = {
+        name: 'whitespace',
+        rules: getWhitespaceRules(grammar)
+    };
+
+    const commentState: State = {
+        name: 'comment',
+        rules: getCommentRules(grammar)
+    };
+
+    // order the following additional rules, to prevent
+    // comment sequences being classified as symbols
+
+    // add include for the whitespace state
+    initialState.rules.push(whitespaceState);
+
+    // add operator & symbol case handling
+    initialState.rules.push({
+        regex: '@symbols',
+        action: [
+            {
+                guard: '@operators',
+                action: { token: 'operator' }
+            },
+            // by default, leave the symbol alone
+            {
+                guard: '@default',
+                action: { token: '' }
+            }
+        ]
+    });
+
+    return [
+        initialState,
+        whitespaceState,
+        commentState
+    ];
+}
+
+/**
+ * Pretty prints a monarch grammar into it's concrete form, suitable for writing to a file
+ * @param monarchGrammar Grammar to pretty print
+ * @returns Monarch grammar in concrete form, suitable for writing to a file
+ */
+function prettyPrint(monarchGrammar: MonarchGrammar): string {
+    const name = monarchGrammar.languageDefinition.name;
+    return ([
+        `// Monarch syntax highlighting for the ${name} language.`,
+        `export const language${name} = {\n`,
+
+        // add language definitions
+        prettyPrintLangDef(monarchGrammar.languageDefinition),
+
+        // add folding
+        '\tfolding: {',
+        '\t\tmarkers: {',
+        '\t\t\tstart: new RegExp(\'^\\s*//\\s*#?region\\b\'),',
+        '\t\t\tend: new RegExp(\'^\\s*//\\s*#?endregion\\b\')',
+        '\t\t}',
+        '\t},\n',
```

Comment:
Suggestion: I don't believe we need folding to be available at the syntax level. We already provide a folding provider in the language server part of Langium.

Reply:
I can omit this in the next commit then.